src-java/nbworker-topology/nbworker-storm-topology/src/main/java/org/openkilda/wfm/topology/nbworker/validators/PathValidator.java

```diff
@@ -0,0 +1,456 @@
+            errors.add(getForwardBandwidthErrorMessage(inputData, forward.get().getAvailableBandwidth()));
+        }
+
+        Optional<Isl> reverse = repositoryData.getIslRepository().findByEndpoints(
+                inputData.getSegment().getDestSwitchId(), inputData.getSegment().getDestPort(),
+                inputData.getSegment().getSrcSwitchId(), inputData.getSegment().getSrcPort());
+        if (!reverse.isPresent()) {
+            return Collections.singleton(getNoForwardIslError(inputData));
+        }
+        if (getReverseBandwidthWithReusableResources(reverse.get(), inputData) < inputData.getPath().getBandwidth()) {
+            errors.add(getReverseBandwidthErrorMessage(inputData, reverse.get().getAvailableBandwidth()));
+        }
+
+        return errors;
+    }
+
+    private long getForwardBandwidthWithReusableResources(Isl isl, InputData inputData) {
+        return getBandwidthWithReusableResources(inputData, isl,
+                pathSegment -> inputData.getSegment().getSrcSwitchId().equals(pathSegment.getSrcSwitchId())
+                        && inputData.getSegment().getDestSwitchId().equals(pathSegment.getDestSwitchId())
+                        && inputData.getSegment().getSrcPort().equals(pathSegment.getSrcPort())
+                        && inputData.getSegment().getDestPort().equals(pathSegment.getDestPort()));
+    }
+
+    private long getReverseBandwidthWithReusableResources(Isl isl, InputData inputData) {
+        return getBandwidthWithReusableResources(inputData, isl,
+                pathSegment -> inputData.getSegment().getSrcSwitchId().equals(pathSegment.getDestSwitchId())
+                        && inputData.getSegment().getDestSwitchId().equals(pathSegment.getSrcSwitchId())
+                        && inputData.getSegment().getSrcPort().equals(pathSegment.getDestPort())
+                        && inputData.getSegment().getDestPort().equals(pathSegment.getSrcPort()));
+    }
+
+    private long getBandwidthWithReusableResources(InputData inputData, Isl isl,
+                                                   Predicate<PathSegment> pathSegmentPredicate) {
+        if (inputData.getPath().getReuseFlowResources() != null
+                && !inputData.getPath().getReuseFlowResources().isEmpty()) {
+
+            Optional<Flow> flow = flowRepository.findById(inputData.getPath().getReuseFlowResources());
+
+            Optional<PathSegment> segment = flow.flatMap(value -> value.getPaths().stream()
+                    .map(FlowPath::getSegments)
+                    .flatMap(List::stream)
+                    .filter(pathSegmentPredicate)
+                    .findAny());
+
+            return segment.map(s -> isl.getAvailableBandwidth() + s.getBandwidth())
+                    .orElseGet(isl::getAvailableBandwidth);
+        }
+
+        return isl.getAvailableBandwidth();
+    }
+
+    private Set<String> validateEncapsulationType(InputData inputData, RepositoryData repositoryData) {
```

Comment:
Shouldn't you also validate encapsulation type on the intermediate switches?

Reply:
The validations are executed against each segment on a path. So it validates the encapsulation type on all switches on a path. Or am I missing something? what do you by intermediate switches?